contracts/MarginPool.sol

```diff
@@ -3,10 +3,135 @@
+        );
+
+        _;
+    }
+
+    /**
+     * @notice fallback function to reveive ETH
+     */
+    receive() external payable {}
+
+    /**
+     * @notice transfers asset from user to pool
+     * @dev all tokens are scaled to have 1e18 precision in contracts,
+     *      so amounts are scaled down to native token decimals using _calcTransferAmount().
+     *      If _asset equal to WETH address, transfer WETH from Controller address to pool
+     * @param _asset address of asset to transfer
+     * @param _user address of user to transfer assets from
+     * @param _amount amount of token to transfer from _user, scaled to 1e18 of precision
+     * @return true if successful transfer
+     */
+    function transferToPool(
+        address _asset,
+        address _user,
+        uint256 _amount
+    ) external onlyController returns (bool) {
+        require(_amount > 0, "MarginPool: transferToPool amount is below 0");
+
+        // get asset decimals
+        uint8 assetDecimal = ERC20Interface(_asset).decimals();
+        // scale amount
+        uint256 val = _calcTransferAmount(_amount, assetDecimal);
+
+        bool success;
+
+        /// check if asset is WETH or not, if WETH transfer the val from Controller address
+        if (_asset == address(WETH)) {
+            // tranfer WETH from controller to pool
+            success = ERC20Interface(_asset).transferFrom(msg.sender, address(this), val);
+        } else {
+            // transfer val from _user to pool
+            success = ERC20Interface(_asset).transferFrom(_user, address(this), val);
+        }
+
+        return success;
+    }
+
+    /**
+     * @notice transfers asset from pool to user
+     * @dev all tokens are scaled to have 1e18 precision in contracts,
+     *      so amounts are scaled down to native token decimals using _calcTransferAmount()
+     * @param _asset address of asset to transfer
+     * @param _user address of user to transfer assets to
+     * @param _amount amount of token to transfer to _user, scaled to 1e18 of precision
+     * @return true if successful transfer
+     */
+    function transferToUser(
+        address _asset,
+        address payable _user,
+        uint256 _amount
+    ) external onlyController returns (bool) {
+        require(_amount > 0, "MarginPool: transferToUser amount is below 0");
+
+        // get asset decimals
+        uint8 assetDecimal = ERC20Interface(_asset).decimals();
+        // scale amount
+        uint256 val = _calcTransferAmount(_amount, assetDecimal);
+
+        // check if asset is WETH or not, if WETH transfer the val from pool to Controller address
+        if (_asset == address(WETH)) {
+            // unwrap WETH
+            WETH.withdraw(val);
+            // transfer ETH to user
+            _user.transfer(val);
+
+            return true;
+        } else {
+            // transfer asset val from Pool to _user
+            return ERC20Interface(_asset).transfer(_user, val);
```

Comment:
fail if this doesn't happen

Reply:
yeah there are require statement in the WETH contract, also we can add require statement in the controller level to check that pool always return true. IMO make sense to add those return checks in the controller level.

Reply:
why not just fail here instead of another check at the controller level?

Reply:
that way we know the controller will not have to worry about another check there right ?

Reply:
I don't know, thought from consistency perspective, all other module return result and Controller make the check.

Reply:
IMO not necessary, making the check from Controller also make sense.